### DISCUSSION TITLE
Can not generate correct sha256 key for webhook.

### DISCUSSION THREAD

**Discussion by flojo007** [community/community#28606] [New to GitHub]
Hello!
When I try to generate the sha256 key, I always get a different one than the one included in the headers. How would I fix this?
Here is my code for the generation:
`app.use(bodyParser.json());`
`app.post('/github-push', async function (req, res) {`
   `console.log("sha256=" + crypto.createHmac('sha256', process.env.GITHUB_PUSH_KEY).update("json").digest("hex")) // TRY 1`
    `console.log("sha256=" + crypto.createHmac('sha256', process.env.GITHUB_PUSH_KEY).update(req + "").digest("hex")) //TRY 2`
    `console.log("sha256=" + crypto.createHmac('sha256', process.env.GITHUB_PUSH_KEY).update(req.read + "").digest("hex")) //TRY 3`
`});`

Thanks a lot, flojo

**Comment by DevAgnihotri** [comment]
Hey,
The issue is that you're not hashing the correct request payload. GitHub generates the SHA-256 signature based on the **raw request body**, not a stringified or modified version of it.  

Use `bodyParser.raw()` instead of `bodyParser.json()` and make sure you're hashing the **exact** body received:  

```javascript
const crypto = require("crypto");
const express = require("express");
const bodyParser = require("body-parser");

const app = express();

// Use raw body parser for correct hashing
app.use(bodyParser.raw({ type: "application/json" }));

app.post("/github-push", (req, res) => {
  const secret = process.env.GITHUB_PUSH_KEY;
  const payload = req.body; // Raw buffer

  const signature = "sha256=" + crypto.createHmac("sha256", secret).update(payload).digest("hex");

  console.log(signature);

  res.status(200).send("Received");
});

app.listen(3000, () => console.log("Listening on port 3000"));
```

- `bodyParser.json()` parses the JSON, which changes the formatting and results in a different hash.  
- `bodyParser.raw({ type: "application/json" })` ensures that you hash the exact raw bytes GitHub used when computing its signature.  

Now, your SHA-256 hash should match the one in GitHub’s headers.
Reach me out if there are any further issues with it.
